stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -204,9 +204,9 @@ inline auto reduce_sum(Vec&& vmapped, int grainsize, std::ostream* msgs,
 
 #ifdef STAN_THREADS
   return internal::reduce_sum_impl<ReduceFunction, void, return_type, Vec,
-                                   Args...>()(std::forward<Vec>(vmapped), true,
-                                              grainsize, msgs,
-                                              std::forward<Args>(args)...);
+                                   ref_type_t<Args&&>...>()(
+      std::forward<Vec>(vmapped), true, grainsize, msgs,
+      std::forward<Args>(args)...);
```

Comment:
So this is using the cast-arguments-once rule to make the `to_ref` happen that evaluates the input arguments if necessary.

What is the difference between what is here and this (so doing the `to_ref` explicitly):
```cpp
return internal::reduce_sum_impl<ReduceFunction, void, return_type, Vec,
  ref_type_t<Args&&>...>()(std::forward<Vec>(vmapped), true,
                           grainsize, msgs,
                           std::forward<ref_tupe_t<Args&&>>(to_ref(std::forward<Args>(args)))...);
```

Is this just more code for the same thing?

Reply:
Yep, `to_ref` is just conversion to `ref_type_t`.

Reply:
Question, if we have an argument `T&& x` that has type `const std::vector<double>&`

Then is it true that:

```
std::is_same<ref_type_t<T&&>, const std::vector<double> &>
```

So that this makes no copy:

```
ref_type_t<T&&> a(deep_copy_vars(x))
```

I was thinking and I think the reason we did `decltype(auto)` there was to make sure `deep_copy` returned references when the input was references so that the only thing that would be saved in the tuple for non-var arguments were references to things we know exist (and rvalues are fine cause how the function is called they won't get deleted).

Now the return type is T, but is there some required copy elison or something that makes sure a copy doesn't happen (a type `T` get constructed by copying the input x which is then returned?)

Reply:
> Then is it true that:

Yes.

> So that this makes no copy:

Yes.

> Now the return type is T, but is there some required copy elison or something that makes sure a copy doesn't happen (a type T get constructed by copying the input x which is then returned?)

Yes, we rely on the compiler to do RVO.

Reply:
Makes sense, thanks!